gaphor/ui/css_function_completion_provider.py

```diff
@@ -0,0 +1,100 @@
+from gi.repository import Gio, GObject, Gtk, GtkSource
+
+
+class CssFunctionCompletionProvider(GObject.GObject, GtkSource.CompletionProvider):
```

Comment:
The code for this class is almost a copy of `CssPropertyCompletionProvider`. I think it's pretty simple to create a simple parameterizable version of a completion provider. Then both completion provider can be in the same file (`gaphor/ui/csscompletion.py`).

Reply:
Yes, it is, right now, but but GtkSourceView is built to support multiple active completion providers, and we can modify these so they differ from each other.
Example it could append `()` after the function name when inserted. It could append a `:` after the attribute name when inserted. 

They also have different [`do_get_title`](https://gnome.pages.gitlab.gnome.org/gtksourceview/gtksourceview5/vfunc.CompletionProvider.get_title.html) and [`do_get_priority`](https://gnome.pages.gitlab.gnome.org/gtksourceview/gtksourceview5/vfunc.CompletionProvider.get_priority.html). The title isn't used by GtkSourceView today, but in future versions it might be. The priority can be used for ordering, example so we can have attributes ordered before functions.

Later, we can add completion providers for colors (they will have a different priority too), and for units for which we might to present when a number is entered, so it is good to keep completions for different things in different completion providers.

Reply:
I like the idea of appending `()` and `:`.

> The priority can be used for ordering, example so we can have attributes ordered before functions.

Thinks like title and priority can be past as constructor parameters to a generic base class, where we handle the common boilerplate.

> Later, we can add completion providers for colors (they will have a different priority too), and for units for which we might to present when a number is entered, so it is good to keep completions for different things in different completion providers.

I see your point. Indeed completers could defer. However, I would not focus to much on later :wink: .

Reply:
> I like the idea of appending () and :.

The appending of characters upon insertion would be done in the [`do_activate`](https://gnome.pages.gitlab.gnome.org/gtksourceview/gtksourceview5/vfunc.CompletionProvider.activate.html) method.

Different providers could have different implementations for the [`is_trigger`](https://gnome.pages.gitlab.gnome.org/gtksourceview/gtksourceview5/vfunc.CompletionProvider.is_trigger.html) method, e.g. to show units (such as `px`) when the user types a digit.